Allow env spec plugins to enable/disable autodetection

soapy1 opened the pull request:
### Description

This PR introduces a way for environment spec plugins to declare that they should not be automatically detected. So, these plugins would only be accessible to users if they explicitly request the plugin by cli arg or configuration.

For example, consider the [environment.yml spec plugin](https://github.com/conda/conda/blob/main/conda/env/specs/yaml_file.py). One could implement a new plugin based on it, with autodetection for their new plugin disabled:
```
class YamlNoAutoDetect(YamlFileSpec):
   detection_supported = False

@hookimpl
def conda_environment_specifiers():
    yield CondaEnvironmentSpecifier(
        name="secret-environment.yml",
        environment_spec=YamlNoAutoDetect,
    )
```

This plugin will never be selected if not explicitly requested. In order to use the `YamlNoAutoDetect` plugin, users must run a command like:

```
$ conda env create --file special.yml --env-spec secret-environment.yml
```

When no plugin is detected, conda will produce an error calling out plugins with autodetection disabled:

```
$ conda env create --file tests/env/support/simple.yml             

EnvironmentSpecPluginNotDetected: Environment at tests/env/support/simple.yml is not readable by any installed environment specifier plugins.

Available plugins: 
    - binstar
    - requirements.txt

Found compatible plugins but they must be explicitly selected.
Request conda to use these plugins by providing
the cli argument `--environment-spec PLUGIN_NAME`:

    - secret-environment.yml
 ```

depends on the changes to the plugin manager introduced in https://github.com/conda/conda/pull/14877

### Checklist - did you ...

<!-- If any of the following items aren't relevant to your contribution,
     please either tick them or use ~strikethrough~ so we know you've gone
     through the checklist. -->

- [x] Add a file to the `news` directory ([using the template](https://github.com/conda/conda/blob/main/news/TEMPLATE)) for the next release's release notes?
     <!-- All "significant" changes should get an entry:
            - user-facing changes or enhancements
            - bug fixes
            - deprecations
            - documentation updates
            - etc -->
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

soapy1 committed:
Allow env spec plugins to enable/disable autodetection

soapy1 committed:
Add news

soapy1 commented on conda/plugins/manager.py:
```diff
@@ -508,39 +508,98 @@ def get_config(self, data) -> PluginConfig:
-                    hook.name,
-                )
-                found.append(hook)
+            if hook.enable_autodetection:
+                log.debug("EnvironmentSpec hook: checking %s", hook.name)
+                if hook.environment_spec(source).can_handle():
+                    log.debug(
+                        "EnvironmentSpec hook: %s can be %s",
+                        source,
+                        hook.name,
+                    )
+                    found.append(hook)
+                else:
+                    log.debug(
+                        "EnvironmentSpec hook: %s can NOT be handled by %s",
+                        source,
+                        hook.name,
+                    )
```
I would say that autodetection and the ability for a plugin to parse an environment spec are seperate ideas. Both of these ideas need to come together eventually for conda to decide if an environment spec plugin is going to be used. However, `can_handle` is about deteriming if a plugin is able to read/parse a given environment spec. And `enable_autodetection` is about a plugin author being able to opt out of the plugin autodetection process.

imo the whole auto detection process is the responsibility of the plugin manager. So, I think the most appropriate place for a plugin author to be able to set this information is in the bit where the plugin get's registered. 

I'm definitely open to suggestions here. But, I would strongly recommend that the mechanism for opting in-to or out-of autodetection is very explicit. Like, it would be confusing if opting out of autodetection was done by having `can_handle` return a `None` value opposed to a `True`/`False` value.

soapy1 replied:
Lets go for adding a `detection_supported` function opposed to a new type. Pushed up the change.

soapy1 committed:
Move plugin autodetection into plugin implementation

soapy1 committed:
Note which plugins have disabled autodetect

soapy1 committed:
Move binstar env spec deprecation to it's __init__ function

Annotating the whole class makes passing refrences to the
class awkward, since it causes 'BinstarSpec' to be a function
opposed to a class. This causes things like CondaEnvironmentSpecifier
to have the wrong type for it's environment_spec field.

soapy1 committed:
Remove duplicate tests

These test cases are already covered in the test_env_specs module.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jaimergp commented on conda/exceptions.py:
```diff
@@ -1297,16 +1297,24 @@ def __init__(self, msg: str, *args, **kwargs):
+            Environment at {name} is not readable by any installed environment specifier plugins.
 
-            Available plugins: {dashlist(plugin_names, 4)}
+Available plugins: {dashlist(plugin_names, 4)}
```
```suggestion
            Available plugins: {dashlist(plugin_names, 4)}
```

jaimergp commented on conda/exceptions.py:
```diff
@@ -1297,16 +1297,24 @@ def __init__(self, msg: str, *args, **kwargs):
             """
         )
+        if len(autodetect_disabled_plugins) > 0:
+            msg += f"\nAvailable plugins with autodetection disabled. Request conda to use these plugins by providing the cli argument `--environment-spec PLUGIN_NAME`: {dashlist(autodetect_disabled_plugins, 4)}"
```
```suggestion
            msg += dals(
                f"""
                Found compatible plugins but they must be explicitly selected.
                Request conda to use these plugins by providing
                the cli argument `--environment-spec PLUGIN_NAME`:
            
                {dashlist(autodetect_disabled_plugins, 4)}"
                """
            )
```

jaimergp commented on conda/exceptions.py:
```diff
@@ -1297,16 +1297,24 @@ def __init__(self, msg: str, *args, **kwargs):
+            Environment at {name} is not readable by any installed environment specifier plugins.
 
-            Available plugins: {dashlist(plugin_names, 4)}
+Available plugins: {dashlist(plugin_names, 4)}
```
What's with the indentation here?

soapy1 replied:
oh, I see actually this can be fixed a better way by setting the right indentation on `dashlist`

soapy1 committed:
Better error message for listing autodetection disabled env spec plugins

Co-authored-by: jaimergp <jaimergp@users.noreply.github.com>

soapy1 commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

soapy1 commented on conda/plugins/types.py:
```diff
@@ -420,6 +420,21 @@ class EnvironmentSpecBase(ABC):
     Base class for all env specs.
     """
 
+    @classmethod
```
Ya, a class attribute sounds good to me. I think the determining if a plugin has autodetect enabled should be done without having to instantiate the object. This is just a setting that a plugin author can set as they see fit. It should not have any extra logic associated with it.

soapy1 committed:
Env plugin detection should be a class attribute

This is just a setting that a plugin author can set as they see
fit. It should not have any extra logic associated with it.

jaimergp commented on conda/env/specs/binstar.py:
```diff
@@ -36,6 +35,7 @@ class BinstarSpec(EnvironmentSpecBase):
 
     msg = None
 
+    @deprecated("24.7", "25.9")
```
Why this change? 🤔

jaimergp commented on conda/exceptions.py:
```diff
@@ -1297,16 +1297,31 @@ def __init__(self, msg: str, *args, **kwargs):
+        self,
+        name: str,
+        plugin_names: Iterable[str],
+        autodetect_disabled_plugins: Iterable[str] = [],
```
```suggestion
        autodetect_disabled_plugins: Iterable[str] = (),
```

Immutable is better.

jaimergp commented on conda/plugins/types.py:
```diff
@@ -420,6 +420,13 @@ class EnvironmentSpecBase(ABC):
+    # If set to False, the only way to use the plugin will be through explicitly
+    # requesting it as a cli argument or setting in .condarc. By default,
+    # autodetection is enabled.
+    detection_supported: bool = True
```
Probably a good idea to mark it as a [ClassVar](https://docs.python.org/3/library/typing.html#typing.ClassVar)

```suggestion
    detection_supported: ClassVar[bool] = True
```

jaimergp commented on news/14914-env-spec-plugins-can-opt-out-of-autodetection:
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* Allow env spec plugins to enable/disable auto detection. (#14914)
```
```suggestion
* Allow env spec plugins to opt-out of auto detection. (#14914)
```

jaimergp commented on conda/exceptions.py:
```diff
@@ -1297,16 +1297,31 @@ def __init__(self, msg: str, *args, **kwargs):
+                """
+                Found compatible plugins but they must be explicitly selected.
+                Request conda to use these plugins by providing
+                the cli argument `--environment-spec PLUGIN_NAME`:
```
Ami I missing something or is this argument not implemented yet?

soapy1 replied:
This was implemented in another PR. If you pull this one down and try it out it should all work as described

jaimergp reviewed (commented):
Looking good, but I think we are missing `--environment-spec` somewhere? Or is this supposed to come in a follow up PR?

soapy1 committed:
Fix some types

Co-authored-by: jaimergp <jaimergp@users.noreply.github.com>

jaimergp commented:
@jezdez, given your previous comments in this PR, can you take another look?
